src/Databases/TablesLoader.h

```diff
@@ -0,0 +1,112 @@
+#pragma once
+#include <Core/Types.h>
+#include <Core/QualifiedTableName.h>
+#include <Parsers/IAST_fwd.h>
+#include <Interpreters/Context_fwd.h>
+#include <Common/ThreadPool.h>
+#include <Common/Stopwatch.h>
+#include <map>
+#include <unordered_map>
+#include <unordered_set>
+#include <mutex>
+
+namespace Poco
+{
+    class Logger;
+}
+
+class AtomicStopwatch;
+
+namespace DB
+{
+
+void logAboutProgress(Poco::Logger * log, size_t processed, size_t total, AtomicStopwatch & watch);
+
+
+class IDatabase;
+using DatabasePtr = std::shared_ptr<IDatabase>;
+
+struct ParsedTableMetadata
+{
+    String path;
+    ASTPtr ast;
+};
+
+using ParsedMetadata = std::map<QualifiedTableName, ParsedTableMetadata>;
+using TableNames = std::vector<QualifiedTableName>;
+
+struct DependenciesInfo
+{
+    /// How many dependencies this table have
+    size_t dependencies_count = 0;
+    /// List of tables/dictionaries which depend on this table/dictionary
+    TableNames dependent_database_objects;
```

Comment:
Dependency_kind (enum) may also be useful sometimes (could be exposed in system.tables for introspection / schema analise) but can be done later